src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
+        final Map<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> unsortedHistogramsAtRGLevel = filteredReads.mapToPair(read -> step1ReadTraversal(read, header))
+                                                                                                                                    .groupByKey()
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step2ConvertToList)
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step3ConvertToHistogram)
+                                                                                                                                    .collectAsMap();
+
+        final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtRGLevel = new TreeMap<>(groupInfoComparator);
+        histogramsAtRGLevel.putAll(unsortedHistogramsAtRGLevel);
+
+        LinkedList<SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>>> listOfStats = new LinkedList<>();
+
+        if (accumLevels.contains(MetricAccumulationLevel.READ_GROUP)) {
+            listOfStats.addFirst(histogramsAtRGLevel);
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.LIBRARY)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.LIBRARY));
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.SAMPLE)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.SAMPLE));
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.ALL_READS)) {
+            listOfStats.addFirst(levelUp(histogramsAtRGLevel, MetricAccumulationLevel.ALL_READS));
+        }
 
-        collectMedianAndMAD(lengthRDD);
+        int i=0;
+        for(final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> map : listOfStats){
+            histogramsAndMetrics.add( convertSortedMapToHTSHistogram(map, histogramMADTolerance) );
+            for(final GroupInfo groupInfo : map.keySet()){ // auxiliary map
+                String name = "";
+                switch (groupInfo.level){
+                    case ALL_READS:
+                        name = ALL_READS_KEY_MACRO;
+                        break;
+                    case SAMPLE:
+                        name = groupInfo.sample;
+                        break;
+                    case LIBRARY:
+                        name = groupInfo.library;
+                        break;
+                    case READ_GROUP:
+                        name = groupInfo.readGroup;
+                        break;
+                }
+                mapFromGroupNameToIndexInList.put(name, i++);
```

Comment:
What happens if a name is not unique across levels -- ie a sample name is the same as a library name, or read group name? 

Could this index map be keyed on GroupInfo instead of a String?